Report availability of Google and Huawei Mobile Services

nategraf opened the pull request:
### Description

We plan to use Google Drive on Android as a component in encrypted cloud backup for account keys,
but in order to use Drive, the user needs to have a Google account and Google Mobile Services (i.e. Google Play Services)
installed on their device. Some Android devices, especially those produced by Huawei since 2019, are
shipped without Google Mobile Services, and the user on these devices is unlikely to have a Google
account.

In order to determine how many Valora users are on Android devices without Google Mobile Services,
this PR introduced functionality to check whether Google Mobile Services are available on startup.
This information is recorded in the state tree and sent in an event to Segment. Sending the event
to Segment will allow us to determine what portion of users will be able to utilize any new feature,
including cloud backup, that will rely on Google Mobile Services.

Additionally, this PR adds a check for Huawei Mobile Services, which is an alternative to Google Mobile Services provided by Huawei on their devices (since they cannot ship with Google integration either in mainland China or, since 2019, abroad)

### Other changes

* Removed unused permission function `requestPhoneStatePermission `

### Tested

* Ran the app and observed the logs for the expected events.

### Related issues

- Related: https://github.com/celo-org/wallet/issues/846

jeanregisser commented on packages/mobile/src/app/actions.ts:
```diff
@@ -33,6 +33,7 @@ export enum Actions {
   APP_MOUNTED = 'APP/APP_MOUNTED',
   APP_UNMOUNTED = 'APP/APP_UNMOUNTED',
   VERIFICATION_MIGRATION_RAN = 'APP/VERIFICATION_MIGRATION_RAN',
+  SET_GOOGLE_PLAY_SERVICES_AVAILABILITY = 'APP/SET_GOOGLE_PLAY_SERVICES_AVAILABILITY',
```
Though the current code base is not a great example of this, moving forward we want to model actions as events instead of setters. See why https://redux.js.org/style-guide/style-guide#model-actions-as-events-not-setters

```suggestion
  GOOGLE_PLAY_SERVICES_AVAILABILITY_CHECKED = 'APP/GOOGLE_PLAY_SERVICES_AVAILABILITY_CHECKED',
```

nategraf replied:
Thanks, fixed in 0128da9886fedc94b4bd4204011f986c779130af

jeanregisser commented on packages/mobile/src/app/actions.ts:
```diff
@@ -130,6 +131,11 @@ export interface VerificationMigrationRanAction {
   now: number
 }
 
+export interface SetGooglePlayServicesAvailability {
```
```suggestion
export interface GooglePlayServicesAvailabilityChecked {
```

jeanregisser commented on packages/mobile/src/app/actions.ts:
```diff
@@ -262,3 +269,10 @@ export const verificationMigrationRan = (
   now: Date.now(),
 })
+
+export const setGooglePlayServicesAvailability = (
```
```suggestion
export const googlePlayServicesAvailabilityChecked = (
```

jeanregisser reviewed (commented):
Looks good!
We can simplify the implementation by using [`react-native-device-info` `hasGms()` call](https://github.com/react-native-device-info/react-native-device-info#hasGms), which [does the native part already](https://github.com/react-native-device-info/react-native-device-info/blob/0a5c21c9a544381577323e48e49cbbbe3bd8c946/android/src/main/java/com/learnium/RNDeviceInfo/RNDeviceModule.java#L503). In that case the redux changes are not even needed.

jeanregisser commented on packages/mobile/src/app/saga.ts:
```diff
@@ -96,6 +102,37 @@ export function* appVersionSaga() {
+// Check the availability of Google Play Services. Log and report the result to.
+// Note: On iOS, this will always report a status of "IOS" and available = false.
+export function* checkGooglePlayServicesSaga() {
+  let result: GooglePlayServicesAvailability
```
Looks like we should just skip this for iOS.

nategraf replied:
Added a check to simply skip on iOS in 4c3a96e6fafaa559f6ef52608aa2114f8d293c96

jeanregisser commented on packages/mobile/src/utils/googleServices.android.ts:
```diff
@@ -0,0 +1,34 @@
+import ReactNative from 'react-native'
+
+const { GooglePlayServicesAvailabilityModule } = ReactNative.NativeModules
+
+export enum GooglePlayServicesAvailability {
+  IOS = -1,
+  SUCCESS = 0,
+  SERVICE_MISSING = 1,
+  SERVICE_VERSION_UPDATE_REQUIRED = 2,
+  SERVICE_DISABLED = 3,
+  SIGN_IN_REQUIRED = 4,
+  INVALID_ACCOUNT = 5,
+  RESOLUTION_REQUIRED = 6,
+  NETWORK_ERROR = 7,
+  INTERNAL_ERROR = 8,
+  SERVICE_INVALID = 9,
+  DEVELOPER_ERROR = 10,
+  LICENSE_CHECK_FAILED = 11,
+  CANCELED = 13,
+  TIMEOUT = 14,
+  INTERRUPTED = 15,
+  API_UNAVAILABLE = 16,
+  SIGN_IN_FAILED = 17,
+  SERVICE_UPDATING = 18,
+  SERVICE_MISSING_PERMISSION = 19,
+  RESTRICTED_PROFILE = 20,
+  RESOLUTION_ACTIVITY_NOT_FOUND = 22,
+  API_DISABLED = 23,
+  API_DISABLED_FOR_CONNECTION = 24,
+}
+
+export async function isGooglePlayServicesAvailable(): Promise<GooglePlayServicesAvailability> {
+  return await GooglePlayServicesAvailabilityModule.isGooglePlayServicesAvailable()
+}
```
In this limited use case, to avoid repeating `GooglePlayServicesAvailability`, we could use a single `googleServices.ts` file.
And use a condition on `Platform.OS === 'ios` in the function.

nategraf replied:
I've removed these files along with the native code

nategraf commented on packages/mobile/android/app/src/main/AndroidManifest.xml:
```diff
@@ -2,7 +2,7 @@
   xmlns:tools="http://schemas.android.com/tools" package="org.celo.mobile" android:versionCode="1" android:versionName="1.0">
 
   <uses-permission android:name="android.permission.INTERNET" />
-  <uses-permission android:name="android.permission.READ_PHONE_STATE" />
```
Oops, I did not think about libraries needing this permission. Fixed in e9d451a015204e473d44ceed6b665f5ac5449664

nategraf commented:
Thanks for your review @jeanregisser! I didn't realize `reactive-native-device-info` already implemented this. I've gone ahead and reworked this PR to use their functions instead of adding new ones. I've also added a check for Huawei Mobile Service since it is available and it would be good to know if that is an option for our users.

jeanregisser reviewed (approved):
Great 🎉
